src/handlers.jl

```diff
@@ -190,17 +190,39 @@ end
 docdict(s::AbstractString) = display_dict(Core.eval(Main, helpmode(devnull, s)))
 
 import Base: is_id_char, is_id_start_char
-function get_token(code, pos)
-    # given a string and a cursor position, find substring to request
-    # help on by:
-    #   1) searching backwards, skipping invalid identifier chars
-    #        ... search forward for end of identifier
-    #   2) search backwards to find the biggest identifier (including .)
-    #   3) if nothing found, do return empty string
-    # TODO: detect operators?
 
+function get_previous_token(code, pos, crossed_parentheses)
+    """
+        get_previous_token(code, pos, crossed_parentheses)
+
+    Given a string and a cursor position, find substring corresponding to previous token.
+    `crossed_parentheses:Int` keeps track of how many parentheses have been crossed.
+    A pair of parentheses yields 0 crossing; a '(' add 1; a ')' subtracts 1.
+
+    Returns `(startpos, endpos, crossed_parentheses, stop)`
+
+    - `startpos` is the start position of the closest potential token before `pos`.
+    - `endpos` is end position if said token is can be valid identifier, or `-1` otherwise
+    - `crossed_parentheses` is the new count for parentheses.
+    - `stop` is true if ';' is hit, denoting the beginning of a clause.
+    """
     startpos = pos
+    separator = false
+    stop = false
     while startpos > firstindex(code)
+        c = code[startpos]
+        if c == '('
+            crossed_parentheses += 1
+            selarator = false
```

Comment:
separator?

Reply:
The "separator" is the hack I referred to in my comments. It basically handles special cases like this:

```
f(..., X, (|Y, ...))
```

On encountering anything not a parenthesis and not an identifier, my algorithm deducts parentheses count. Without the hack, the token returned would be `X` instead of `f`. It is an incredibly bad hack which is definitely not the correct algorithm, but it works in a wide variety of cases and is at least better than the original implementation which can only return the token immediately before cursor.